data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,1255 @@
+        let mut vtt_lowest_absolute = None;
+        let mut vtt_highest_absolute = Priority::default();
+
+        let epsilon = 0.1 / buckets_count;
+        let mut drt_counter = LossyCounter::<u64>::with_epsilon(epsilon);
+        let mut vtt_counter = LossyCounter::<u64>::with_epsilon(epsilon);
+
+        for (
+            epoch,
+            Priorities {
+                drt_highest,
+                drt_lowest,
+                vtt_highest,
+                vtt_lowest,
+            },
+        ) in priorities.enumerate()
+        {
+            // Keep track of the lowest and highest recorded priorities.
+            option_update_if_less_than(&mut drt_lowest_absolute, *drt_lowest);
+            if drt_highest > &drt_highest_absolute {
+                drt_highest_absolute = *drt_highest;
+            }
+            option_update_if_less_than(&mut vtt_lowest_absolute, *vtt_lowest);
+            if vtt_highest > &vtt_highest_absolute {
+                vtt_highest_absolute = *vtt_highest;
+            }
+
+            let epoch = epoch as f64;
+            let drt_lowest = drt_lowest.unwrap_or(*drt_highest).as_f64();
+            let drt_highest_absolute = drt_highest_absolute.as_f64();
+            let drt_lowest_absolute = drt_lowest_absolute
+                .map(Priority::into_inner)
+                .unwrap_or(drt_highest_absolute);
+            let vtt_lowest = vtt_lowest.unwrap_or(*vtt_highest).as_f64();
+            let vtt_highest_absolute = vtt_highest_absolute.as_f64();
+            let vtt_lowest_absolute = vtt_lowest_absolute
+                .map(Priority::into_inner)
+                .unwrap_or(vtt_highest_absolute);
+
+            // This calculates the bucket in which the lowest values should be inserted, by applying
+            // a rolling "compander" mechanism. That is, we keep track of the absolute range of
+            // lowest to highest priority values for all the previous epochs, to then compare the
+            // range in each epoch and map one range to another to find its relative position.
+            let drt_lowest_compared_to_absolute = drt_lowest / drt_lowest_absolute;
+            let drt_range_absolute = drt_highest_absolute / drt_lowest_absolute;
+            let drt_bucket_index =
+                drt_lowest_compared_to_absolute / drt_range_absolute * buckets_count;
+            let vtt_lowest_compared_to_absolute = vtt_lowest / vtt_lowest_absolute;
+            let vtt_range_absolute = vtt_highest_absolute / vtt_lowest_absolute;
+            let vtt_bucket_index =
+                vtt_lowest_compared_to_absolute / vtt_range_absolute * buckets_count;
+
+            // For a perfect calculation, all values lower than the lowest bucket index
+            // (representing the lowest fee should be inserted. However, we can get a good enough
+            // approximation while saving CPU and memory by inserting only the 10% closest values
+            // from below.
+            if epoch > 0.0 {
```

Comment:
This was introduced back then to prevent comparing with uninitialized absolutes and averages, which could lead to overflows and stuff being inserted into the wrong buckets. But you are right, this check and the `epoch` enumerator is now redundant altogether.